packages/grid/x-data-grid/src/components/menu/columnMenu/GridColumnMenu.tsx

```diff
@@ -1,47 +1,101 @@
 import * as React from 'react';
 import PropTypes from 'prop-types';
-import { GridColumnMenuContainer } from './GridColumnMenuContainer';
+import { GridColumnMenuTypes, GridColumnMenuValue } from '../../../hooks/features/columnMenu';
 import { GridColumnMenuProps } from './GridColumnMenuProps';
-import { GridColumnsMenuItem } from './GridColumnsMenuItem';
-import { GridFilterMenuItem } from './GridFilterMenuItem';
-import { HideGridColMenuItem } from './HideGridColMenuItem';
-import { SortGridMenuItems } from './SortGridMenuItems';
 import { useGridApiContext } from '../../../hooks/utils/useGridApiContext';
 
-const GridColumnMenu = React.forwardRef<HTMLUListElement, GridColumnMenuProps>(
-  function GridColumnMenu(props: GridColumnMenuProps, ref) {
-    const { hideMenu, currentColumn } = props;
-    const apiRef = useGridApiContext();
-
-    const defaultButtons = [
-      <SortGridMenuItems onClick={hideMenu} column={currentColumn!} />, // TODO update types to allow `onClick` and `column` to be optional
-      <GridFilterMenuItem onClick={hideMenu} column={currentColumn!} />,
-      <HideGridColMenuItem onClick={hideMenu} column={currentColumn!} />,
-      <GridColumnsMenuItem onClick={hideMenu} column={currentColumn!} />,
-    ];
-
-    const preProcessedButtons = apiRef.current.unstable_applyPipeProcessors(
-      'columnMenu',
-      defaultButtons,
-      currentColumn,
-    );
-
-    return (
-      <GridColumnMenuContainer ref={ref} {...props}>
-        {preProcessedButtons.map((button: any, index: number) =>
-          React.cloneElement(button, { key: index, onClick: hideMenu, column: currentColumn }),
-        )}
-      </GridColumnMenuContainer>
-    );
-  },
-);
+const GridColumnMenu = (props: GridColumnMenuProps) => {
+  const {
+    hideMenu,
+    currentColumn,
+    defaultMenuItems,
+    defaultVisibleItems,
+    columnMenuItems: userMenuItems,
+  } = props;
+
+  const apiRef = useGridApiContext();
+
+  const preProcessedValue = apiRef.current.unstable_applyPipeProcessors(
+    'columnMenu',
+    { items: defaultMenuItems, visibleItemKeys: defaultVisibleItems },
+    currentColumn,
+  );
+
+  const extendedColumnMenuItems: GridColumnMenuValue['items'] = React.useMemo(() => {
```

Comment:
Type inference should be enough here :)

```suggestion
  const extendedColumnMenuItems = React.useMemo(() => {
```